stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -0,0 +1,946 @@
+      // finesse the hash value of -0.0 to match +0.0.
+      return 0
+    } else {
+    %if bits <= word_bits:
+      return Int(bitPattern: UInt(bitPattern))
+    %elif bits == 64: # Double -> 32-bit Int
+      return Int(truncatingBitPattern: bitPattern >> 32) ^
+             Int(truncatingBitPattern: bitPattern)
+    %elif word_bits == 32: # Float80 -> 32-bit Int
+      return Int(truncatingBitPattern: UInt(significandBitPattern >> 32)) ^
+             Int(truncatingBitPattern: UInt(significandBitPattern)) ^
+             Int(_representation.signAndExponent)
+    %else: # Float80 -> 64-bit Int
+      return Int(bitPattern: UInt(significandBitPattern)) ^
+             Int(_representation.signAndExponent)
+    %end
+    }
+  }
+}
+
+@_transparent
+extension ${Self} : AbsoluteValuable {
+  /// Returns the absolute value of `x`.
+  @_transparent
+  @warn_unused_result
+  public static func abs(_ x: ${Self}) -> ${Self} {
+    return ${Self}(_bits: Builtin.int_fabs_FPIEEE${bits}(x._value))
+  }
+}
+
+@_transparent
+@warn_unused_result
+public prefix func +(x: ${Self}) -> ${Self} {
+  return x
+}
+
+@_transparent
+@warn_unused_result
+public prefix func -(x: ${Self}) -> ${Self} {
+  return ${Self}(_bits: Builtin.fneg_FPIEEE${bits}(x._value))
+}
+
+//===----------------------------------------------------------------------===//
+// Explicit conversions between types.
+//===----------------------------------------------------------------------===//
+
+// Construction from integers.
+@_transparent
+extension ${Self} {
+% for (srcBits, srcSigned) in allInts():
+%    That = intName(srcBits, srcSigned)
+%    ThatBuiltinName = builtinIntName(srcBits)
+%    sign = 's' if srcSigned else 'u'
+  public init(_ v: ${That}) {
+    _value = Builtin.${sign}itofp_${ThatBuiltinName}_FPIEEE${bits}(v._value)
+  }
+% end
+}
+
+// Construction from other floating point numbers.
+@_transparent
+extension ${Self} {
+% for srcBits in allFloatBits:
+%   That = floatName[srcBits]
+%   if Self != That:
+
+%     if srcBits == 80:
+#if !os(Windows) && (arch(i386) || arch(x86_64))
+%     end
+
+  /// Construct an instance that approximates `other`.
+  public init(_ other: ${That}) {
+%     if srcBits > bits:
+    _value = Builtin.fptrunc_FPIEEE${srcBits}_FPIEEE${bits}(other._value)
+%     else:
+    _value = Builtin.fpext_FPIEEE${srcBits}_FPIEEE${bits}(other._value)
+%     end
+  }
+
+%     if srcBits == 80:
+#endif
+%     end
+
+%   end
+% end
+}
+
+//===----------------------------------------------------------------------===//
+// Standard Operator Table
+//===----------------------------------------------------------------------===//
+
+//  TODO: These should not be necessary, since they're already provided by
+//  <T: FloatingPoint>, but in practice they are currently needed to
+//  disambiguate overloads.  We should find a way to remove them, either by
+//  tweaking the overload resolution rules, or by removing the other
+//  definitions in the standard lib, or both.
```

Comment:
> On May 9, 2016, at 9:58 AM, Dmitri Gribenko <notifications@github.com> wrote:
> 
> In stdlib/public/core/FloatingPointTypes.swift.gyb <https://github.com/apple/swift/pull/2453#discussion_r62533536>:
> 
> > +#endif
> > +%     end
> > +
> > +%   end
> > +% end
> > +}
> > +
> > +//===----------------------------------------------------------------------===//
> > +// Standard Operator Table
> > +//===----------------------------------------------------------------------===//
> > +
> > +//  TODO: These should not be necessary, since they're already provided by
> > +//  <T: FloatingPoint>, but in practice they are currently needed to
> > +//  disambiguate overloads.  We should find a way to remove them, either by
> > +//  tweaking the overload resolution rules, or by removing the other
> > +//  definitions in the standard lib, or both.
> Do you have a bug? We could ask @jopamer <https://github.com/jopamer> to look.
> 
Thanks - if there’s a bug, I’ll be happy to take a look.

- Joe

> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/2453/files/6429eabf749777f5f9f973875760c3848c58a824#r62533536>